src/rust/engine/client/src/options/parse.rs

```diff
@@ -0,0 +1,152 @@
+            = "\"" s:double_quoted_character()* "\"" { s.into_iter().collect() }
+
+        rule double_quoted_character() -> char
+            = quoted_character("\"")
+            / escaped_character()
+
+        rule single_quoted_string() -> String
+            = "'" s:single_quoted_character()* "'" { s.into_iter().collect() }
+
+        rule single_quoted_character() -> char
+            = quoted_character("'")
+            / escaped_character()
+
+        rule quoted_character(quote_char: &'static str) -> char
+            = !(##parse_string_literal(quote_char) / "\\") c:$([_]) { c.chars().next().unwrap() }
+
+        rule escaped_character() -> char
+            = "\\" c:$([_]) { c.chars().next().unwrap() }
+
+        rule add() -> ListEditAction
+            = "+" { ListEditAction::ADD }
+
+        rule remove() -> ListEditAction
+            = "-" { ListEditAction::REMOVE }
+
+        rule action() -> ListEditAction
+            = quiet!{ action:(add() / remove()) { action } }
+            / expected!(
+                "an optional list edit action of '+' indicating `add` or '-' indicating `remove`"
+            )
+
+        // N.B.: The Python list parsing implementation accepts Python tuple literal syntax too.
+
+        rule tuple_start() -> ()
+            = quiet!{ "(" }
+            / expected!("the start of a tuple indicated by '('")
+
+        rule tuple_end() -> ()
+            = quiet!{ ")" }
+            / expected!("the end of a tuple indicated by ')'")
+
+        rule tuple_items() -> Vec<String>
+            = tuple_start() items:string() ** "," ","? tuple_end() { items }
+
+        rule list_start() -> ()
+            = quiet!{ "[" }
+            / expected!("the start of a list indicated by '['")
+
+        rule list_end() -> ()
+            = quiet!{ "]" }
+            / expected!("the end of a list indicated by ']'")
+
+        rule list_items() -> Vec<String>
+            = list_start() items:string() ** "," ","? list_end() { items }
+
+        rule items() -> Vec<String>
+            = tuple_items()
+            / list_items()
+
+        rule list_edit() -> ListEdit<String>
+            = whitespace()* action:action() items:items() whitespace()* {
+                ListEdit { action, items }
+            }
+
+        rule list_edits() -> Vec<ListEdit<String>>
+            = e:list_edit() ** "," ","? { e }
+
+        rule list_replace() -> Vec<ListEdit<String>>
+            = items:items() {
+                vec![ListEdit { action: ListEditAction::REPLACE, items }]
+            }
+
+        rule implicit_add() -> Vec<ListEdit<String>>
+            = !(whitespace() / add() / remove() / tuple_start() / list_start()) item:$([_]+) {
+                vec![ListEdit { action: ListEditAction::ADD, items: vec![item.to_owned()] }]
+            }
+
+        pub(crate) rule string_list_edits() -> Vec<ListEdit<String>>
+            = list_edits() / list_replace() / implicit_add()
+    }
+}
+
+fn format_parse_error(
```

Comment:
This goes through some effort to present a useful error message. For example this shows both the native client error message and the python error message since any failure currently in the native client prior to nailgun session establishment has the pants script falling through to the python client:
```
$ USE_NATIVE_PANTS=1 PANTS_CONFIG_FILES="-['/etc/hosts'],           
         ?(\"/dev/null\")
" ./pants -V
Problem parsing PANTS_CONFIG_FILES string list value:
1:-['/etc/hosts'],
2:         ?("/dev/null")
  ---------^
3:
Expected an optional list edit action of '+' indicating `add` or '-' indicating `remove` at line 2 column 10
                The value cannot be evaluated as a literal expression: IndentationError('unexpected indent', ('<string>', 2, 9, '         ?("/dev/null")\n'))
                Given raw value:
                1: ['/etc/hosts'],
2:          ?("/dev/null")
Traceback (most recent call last):
  File "/home/jsirois/dev/pantsbuild/jsirois-pants/src/python/pants/util/eval.py", line 47, in parse_expression
    parsed_value = eval(val)
  File "<string>", line 2
    ?("/dev/null")
    ^
IndentationError: unexpected indent
...
```

Reply:
Or:
```
$ USE_NATIVE_PANTS=1 PANTS_CONFIG_FILES="-['/etc/hosts'],+(\"/dev/null\"]" ./pants -V
Problem parsing PANTS_CONFIG_FILES string list value:
1:-['/etc/hosts'],+("/dev/null"]
  -----------------------------^
Expected "," or the end of a tuple indicated by ')' at line 1 column 30
The value cannot be evaluated as a literal expression: SyntaxError('unexpected EOF while parsing', ('<string>', 1, 13, '("/dev/null"]'))
Given raw value:
1: ("/dev/null"]
Traceback (most recent call last):
  File "/home/jsirois/dev/pantsbuild/jsirois-pants/src/python/pants/util/eval.py", line 47, in parse_expression
    parsed_value = eval(val)
  File "<string>", line 1
    ("/dev/null"]
                ^
SyntaxError: unexpected EOF while parsing
```